// @neoaz07/nkxfca/src/apis/getUserID.js
"use strict";

const utils = require('../utils');

function formatData(data) {
  return {
    userID: utils.formatID(data.uid.toString()),
    photoUrl: data.photo,
    indexRank: data.index_rank,
    name: data.text,
    isVerified: data.is_verified,
    profileUrl: data.path,
    category: data.category,
    score: data.score,
    type: data.type
  };
}

module.exports = (defaultFuncs, api, ctx) => {
  return async function getUID(link, callback) {
    let resolveFunc = () => {};
    let rejectFunc = () => {};
    const returnPromise = new Promise((resolve, reject) => {
      resolveFunc = resolve;
      rejectFunc = reject;
    });

    if (!callback) {
      callback = (err, result) => {
        if (err) return rejectFunc(err);
        resolveFunc(result);
      };
    }

    if (!link || typeof link !== 'string') {
      const error = { error: "getUID: link parameter must be a non-empty string" };
      utils.error("getUID", error);
      return callback(error);
    }

    // Check if it's a profile URL
    const isProfileUrl = link.match(/\.com/);
    if (!isProfileUrl) {
      // Treat as username/name, use search
      try {
        const form = {
          value: link.toLowerCase(),
          viewer: ctx.userID,
          rsp: "search",
          context: "search",
          path: "/home.php",
          request_id: ctx.clientID || utils.getGUID()
        };

        const res = await defaultFuncs.get("https://www.facebook.com/ajax/typeahead/search.php", ctx.jar, form)
          .then(utils.parseAndCheckLogin(ctx, defaultFuncs));

        if (res.error) {
          throw res;
        }

        if (!res.payload || !res.payload.entries) {
          const error = { 
            error: "getUID: No results found. This may be due to Facebook security restrictions or account checkpoint.",
            details: "Your account may require verification. Please visit facebook.com to verify."
          };
          throw error;
        }

        const data = res.payload.entries;
        
        if (data.length === 0) {
          utils.warn(`getUID: No user found with name "${link}"`);
        }

        callback(null, data.map(formatData));
      } catch (err) {
        if (err.error && typeof err.error === 'string' && err.error.includes('checkpoint')) {
          err.friendlyMessage = "Account checkpoint required - Please verify your account on facebook.com";
        }
        utils.error("getUID", err);
        callback(err);
      }
      return returnPromise;
    }

    // Handle profile URL
    try {
      let uid;
      if (link.includes('profile.php?id=')) {
        uid = link.split('profile.php?id=')[1].split('&')[0];
      } else {
        // For username URLs, fetch the page to get userID
        const res = await defaultFuncs.get(link, ctx.jar)
          .then(utils.parseAndCheckLogin(ctx, defaultFuncs));
        
        const userIDMatch = res.match(/"userID":"(\d+)"/) || res.match(/"id":"(\d+)"/) || res.match(/entity_id["\s:]+(\d+)/);
        if (userIDMatch) {
          uid = userIDMatch[1];
        } else {
          throw new Error("Could not extract user ID from profile URL");
        }
      }

      if (!uid || !/^\d+$/.test(uid)) {
        throw new Error("Invalid user ID extracted");
      }

      callback(null, uid);
    } catch (err) {
      utils.error("getUID", err);
      callback(err);
    }

    return returnPromise;
  };
};
